File: desktop/src/tierLimits.ts
import type { SubscriptionTier } from "./auth";

// =====================================================================
// Tier Limits
//
// SOURCE OF TRUTH: api/core/tier_limits.go (DefaultTierLimits)
//
// Channel config panels and the onboarding wizard read these synchronously
// during render, so we keep a hardcoded mirror of the backend values here
// rather than fetching them asynchronously from GET /tier-limits. Drift
// between this file and the Go source becomes a billing-trust problem.
//
// If you change a number here, you MUST also update:
//   - api/core/tier_limits.go        (the Go map)
//   - api/core/tier_limits_test.go   (the assertion)
//   - myscrollr.com/src/routes/uplink.tsx (the FALLBACK_LIMITS constant,
//     for first-paint before the runtime fetch resolves)
//
// Infinity here corresponds to `null` on the wire (null round-trips
// through JSON; Infinity does not).
// =====================================================================

interface ChannelLimits {
  symbols: number;
  feeds: number;
  customFeeds: number;
  leagues: number;
  fantasy: number;
}

export const TIER_LIMITS: Record<SubscriptionTier, ChannelLimits> = {
  free: { symbols: 5, feeds: 1, customFeeds: 0, leagues: 1, fantasy: 0 },
  uplink: { symbols: 25, feeds: 25, customFeeds: 1, leagues: 8, fantasy: 1 },
  uplink_pro: { symbols: 75, feeds: 100, customFeeds: 3, leagues: 20, fantasy: 3 },
  uplink_ultimate: {
    symbols: Infinity,
    feeds: Infinity,
    customFeeds: 10,
    leagues: Infinity,
    fantasy: 10,
  },
  super_user: {
    symbols: Infinity,
    feeds: Infinity,
    customFeeds: Infinity,
    leagues: Infinity,
    fantasy: Infinity,
  },
};

type LimitKey = keyof ChannelLimits;

/** Get the numeric limit for a tier + channel feature. */
export function getLimit(tier: SubscriptionTier, key: LimitKey): number {
  return TIER_LIMITS[tier][key];
}

/** True when the tier has no cap (Infinity) for the given feature. */
export function isUnlimited(tier: SubscriptionTier, key: LimitKey): boolean {
  return TIER_LIMITS[tier][key] === Infinity;
}

/**
 * Returns `maxItems` for SetupBrowser: a finite number or undefined (unlimited).
 * Passing undefined means SetupBrowser won't enforce any cap.
 */
export function maxItemsForBrowser(
  tier: SubscriptionTier,
  key: LimitKey
): number | undefined {
  const limit = TIER_LIMITS[tier][key];
  return limit === Infinity ? undefined : limit;
}

File: api/core/tier_limits.go
package core

import "github.com/gofiber/fiber/v2"

// ChannelLimits is the per-tier cap for every channel feature the frontend
// lets users configure. A nil pointer means "unlimited" — this lets us
// round-trip cleanly through JSON (where Go's Infinity has no analogue)
// and lets clients treat null as "no cap."
type ChannelLimits struct {
	Symbols     *int `json:"symbols"`
	Feeds       *int `json:"feeds"`
	CustomFeeds *int `json:"custom_feeds"`
	Leagues     *int `json:"leagues"`
	Fantasy     *int `json:"fantasy"`
}

// TierLimitsResponse is the payload of GET /tier-limits.
type TierLimitsResponse struct {
	Tiers map[string]ChannelLimits `json:"tiers"`
}

// DefaultTierLimits is the authoritative source of truth for per-tier caps.
//
// SOURCE OF TRUTH — any change here MUST also propagate to:
//   - desktop/src/tierLimits.ts     (kept in sync manually; synchronous
//     reads required by config panels during render)
//   - myscrollr.com/src/routes/uplink.tsx (`FALLBACK_LIMITS` constant,
//     used only for first-paint while the API response is in flight)
//   - api/core/tier_limits_test.go  (assertion protecting this table
//     from silent edits — run `go test ./core/...` after any change)
//
// Once Sprint 3 wires backend enforcement on POST/PUT /users/me/channels,
// these values directly gate what the DB will accept, so drift is
// unforgiving.
var DefaultTierLimits = map[string]ChannelLimits{
	"free":            {Symbols: intPtr(5), Feeds: intPtr(1), CustomFeeds: intPtr(0), Leagues: intPtr(1), Fantasy: intPtr(0)},
	"uplink":          {Symbols: intPtr(25), Feeds: intPtr(25), CustomFeeds: intPtr(1), Leagues: intPtr(8), Fantasy: intPtr(1)},
	"uplink_pro":      {Symbols: intPtr(75), Feeds: intPtr(100), CustomFeeds: intPtr(3), Leagues: intPtr(20), Fantasy: intPtr(3)},
	"uplink_ultimate": {Symbols: nil, Feeds: nil, CustomFeeds: intPtr(10), Leagues: nil, Fantasy: intPtr(10)},
	"super_user":      {Symbols: nil, Feeds: nil, CustomFeeds: nil, Leagues: nil, Fantasy: nil},
}

// HandleGetTierLimits serves the tier limits map to any caller — clients
// render pricing/comparison UIs from this, and integration tests use it
// to confirm desktop and marketing values agree with the backend.
//
// Unauthenticated on purpose: these numbers are marketing-visible, and
// we want the pricing page to load without a session.
func HandleGetTierLimits(c *fiber.Ctx) error {
	// Short browser + CDN cache. The pricing page fetches this on mount;
	// a 5-minute cache is generous enough to reduce load while still
	// letting us ship a limit change without waiting hours.
	c.Set("Cache-Control", "public, max-age=300")
	return c.JSON(TierLimitsResponse{Tiers: DefaultTierLimits})
}

// intPtr returns a pointer to an int literal — convenience for the table
// above so each row stays readable.
func intPtr(n int) *int {
	return &n
}
